Addition of three new predefined recognizers, improved regex for IN_PAN

devopam opened the pull request:
## Change Description
Added three new recognizers
1. IN_GSTIN
2. ISIN_CODE
3. CFI_CODE

Improved IN_PAN regex to not get spoofed by 0000 pattern

Describe your changes
Added new recognizers to improve detection of PII elements , one India specific and two global.

## Issue reference

This PR fixes issue #XX

## Checklist

- [X] I have reviewed the [contribution guidelines](https://github.com/microsoft/presidio/blob/main/CONTRIBUTING.md)
- [X] I have signed the CLA (if required)
- [X] My code includes unit tests
- [X] All unit tests and lint checks pass locally
- [X] My PR contains documentation updates / additions if required

devopam commented:
hi @omri374 ,
Apologies if I made it bit heavy by adding three recognizers in one go.
One additional library used : pycountry (https://pypi.org/project/pycountry/) , please let me know if their licenses are okay to be included within presidio.

regards
Devopam

omri374 commented:
Hi @devopam, thanks for this! unfortunately the `pycountry` package is LGPL and we cannot have a dependency on it.

omri374 commented:
Is there a simple alternative to it?

devopam commented:
hi @omri374 , yes I will add it neatly in analyzer_utils so that others can use it as well. We will need to maintain all these metadata whenever world events change as such. I am down with cold  & fever :( - that's the delay cause.

omri374 commented:
Absolutely no rush! Hope you feel better soon!

devopam commented:
hi @omri374 , 
I have removed pycountry as desired and added the metadata to country_master.csv file. Added util functions in analyzer_utils to load the required information as well (added one function for future use also). However, I am not convinced of the location of country_master.csv myself - shall we create a new sub-folder at the top level for such data ? 
Please have a look when you can and let me know your feedback on the overall changes.

omri374 commented:
Thanks! I will take a look soon.

omri374 commented on presidio-analyzer/presidio_analyzer/country_master.csv:
```diff
@@ -0,0 +1,255 @@
+﻿English_full_name,English_short_name_using_title_case,FIFA_country_code,International_olympic_committee_country_code,ISO3166-1-Alpha-2,ISO3166-1-Alpha-3,ISO3166-1-Numeric,international licence plate country code,Country_code_top_level_domain,Currency Name,ISO4217-Alpha-3,ISO4217-Numeric,Currency Minor Unit,Capital_City,Recognized_Country_Flag,Dialing_Code,Mobile_Country_Code,Mobile_Network_Code
```
I would suggest to put this file in a new `data` folder.

devopam replied:
yes. i had the same inkling , implemented .

omri374 commented on presidio-analyzer/presidio_analyzer/predefined_recognizers/in_gstin_recognizer.py:
```diff
@@ -0,0 +1,152 @@
+
+    gstin_country_codes_iso3a = ""
+    utils = Utils()
+    countries = utils.get_country_codes(iso_code="ISO3166-1-Alpha-3")
```
I'm trying to think of way to make this more efficient. Creating the `PresidioAnalyzerUtils` object for every recognizer that needs it is inefficient, and would require holding the countries data in memory multiple times.

On one hand, there's the need for this data for some recognizers, on the other hand most recognizers (and users) might not need it. Therefore, we can think of two options to handle this in my view.
1. Instantiate the `utils` in the `AnalyzerEngine` class and pass it to each recognizer. This would allow it to only be instantiated once.
2. Make some of these recognizers optional (which aren't loaded in `load_predefined_recognizers`) and therefore reduce the memory footprint. 

Long term, I think we should think of a mechanism to specify the countries you expect PII to come from. If someone doesn't expect any Australian license plates, it doesn't make sense to load that recognizer for them.

Happy to get your thoughts on this.

devopam replied:
hi @omri374 
I am confused with the best approach here frankly.  #2 is not a good idea as it will defeat the purpose of 'built-in' recognizers readily available as such from a consumer's perspective imho. 
We will need more of such metadata to improve detection beyond pattern matching as we add more industry specific recognizers  to give comprehensive solution offering.
So, how best to design this is bit complex ! 
#1 above seems to be a feasible approach to me but I will need fair bit of help here to implement since I don't want to end up breaking things due to my limited knowledge of the core. 
Shall we discuss offline ? Please suggest

omri374 replied:
Hi @devopam. I agree that (1) is a better approach, also since in parallel we're thinking of introducing a country flag to be able to filter out recognizers that aren't needed.

Adding a new input to the `__init__` of `EntityRecognizer` would not break the API. I would suggest to do that, and instantiate the `PresidioAnalyzerUtils` in the recognizer registry, so that it could be passed to all recognizers during instantiation.

How about taking this constructor:
https://github.com/microsoft/presidio/blob/ea8d830fe445ee150e4ed76db5b4f36904400809/presidio-analyzer/presidio_analyzer/entity_recognizer.py#L35

and adding a new param to it:
```py
    def __init__(
        self,
        supported_entities: List[str],
        name: str = None,
        supported_language: str = "en",
        version: str = "0.0.1",
        context: Optional[List[str]] = None,
        analyzer_utils: Optional[PresidioAnalyzerUtils] = None
    ):
        self.analyzer_utils = analyzer_utils

```

We could inject the analyzer_utils during recognizers instantiation:
https://github.com/microsoft/presidio/blob/ea8d830fe445ee150e4ed76db5b4f36904400809/presidio-analyzer/presidio_analyzer/recognizer_registry.py#L139

WDYT?
Happy to discuss offline too.

omri374 replied:
@devopam how about we have a quick chat on this? If you're interested, let's chat on LinkedIn

devopam replied:
Hi @omri374 ,
Please have a look at the current implementation and advise. Utils is instantiated only in the recognizers using it now viz. in_gstin, isin

omri374 commented on presidio-analyzer/tests/test_isin_recognizer.py:
```diff
@@ -0,0 +1,61 @@
+    return ["ISIN_CODE"]
+
+
+@pytest.mark.parametrize(
```
Great work on testing!

omri374 commented on presidio-analyzer/presidio_analyzer/predefined_recognizers/isin_recognizer.py:
```diff
@@ -0,0 +1,61 @@
+    PATTERNS = [
+        Pattern(
+            "ISIN (Medium)",
+            r"\b[A-Z]{2}[A-Z0-9]{9}\d{1}\b",
```
Consider reducing the score on this regex. It's quite generic (I would suggest 0.01)

omri374 reviewed (commented):
Thanks for the great work! Left some comments, would be happy to discuss.

omri374 commented:
/azp run

devopam commented on presidio-analyzer/presidio_analyzer/analyzer_utils.py:
```diff
@@ -73,3 +101,140 @@ def is_verhoeff_number(input_number: int):
+            raise FileNotFoundError()
+        else:
+            with open(
+                file=self.__country_master_file_path__,
```
incorporated

devopam commented on presidio-analyzer/tests/test_recognizer_registry.py:
```diff
@@ -57,8 +57,8 @@ def test_when_get_recognizers_then_all_recognizers_returned(mock_recognizer_regi
     recognizers = registry.get_recognizers(language="en", all_fields=True)
-    # 1 custom recognizer in english + 24 predefined
-    assert len(recognizers) == 1 + 24
+    # 1 custom recognizer in english + 27 predefined
```
silly me :( . fixed the comment

omri374 commented on presidio-analyzer/presidio_analyzer/analyzer_utils.py:
```diff
@@ -36,13 +52,33 @@ def sanitize_value(text: str, replacement_pairs: List[Tuple[str, str]]) -> str:
         return text
 
+    @staticmethod
+    def get_luhn_mod_n(input_str: str, alphabet="0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ"):
```
```suggestion
    def get_luhn_mod_n(input_str: str, alphabet="0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ") -> bool:
```

omri374 commented on presidio-analyzer/presidio_analyzer/analyzer_utils.py:
```diff
@@ -36,13 +52,33 @@ def sanitize_value(text: str, replacement_pairs: List[Tuple[str, str]]) -> str:
+        ) % n == 0
+
     @staticmethod
     def is_verhoeff_number(input_number: int):
```
```suggestion
    def is_verhoeff_number(input_number: int) -> bool:
```

omri374 commented on presidio-analyzer/presidio_analyzer/analyzer_utils.py:
```diff
@@ -73,3 +109,140 @@ def is_verhoeff_number(input_number: int):
+            # return full country list for given code
+            return self.__get_country_master_full_data__(iso_code=iso_code)
+
+    def get_full_country_information(self, lookup_key: str, lookup_index: str):
```
please define return type (`List[str]`?)

omri374 commented on presidio-analyzer/presidio_analyzer/analyzer_utils.py:
```diff
@@ -73,3 +109,140 @@ def is_verhoeff_number(input_number: int):
+        ISO3166-1-Alpha-2,ISO3166-1-Alpha-3, ISO3166-1-Numeric,
+        International_licence_plate_country_code, Country_code_top_level_domain,
+        Currency_Name, ISO4217-Alpha-3, ISO4217-Numeric, Capital_City, Dialing_Code
+        :return: Dictionary object with additional information enriched from
```
It says it returns a dictionary, but it looks like the code returns a `list`

omri374 commented on presidio-analyzer/presidio_analyzer/entity_recognizer.py:
```diff
@@ -54,11 +54,15 @@ def __init__(
         self.is_loaded = True
 
+        if analyzer_utils is not None:
+            self.analyzer_utils = analyzer_utils
```
class fields should not be optional. I would suggest to remove the `if analyzer_utils is not None` and put the value anyway (even if it's None). Otherwise the `analyzer_utils` field would not always be part of the class.

omri374 commented on presidio-analyzer/presidio_analyzer/pattern_recognizer.py:
```diff
@@ -50,12 +52,13 @@ def __init__(
                 " or with deny list"
             )
-
+        print(supported_entity)
```
Please remove prints

omri374 commented on presidio-analyzer/presidio_analyzer/entity_recognizer.py:
```diff
@@ -39,10 +40,9 @@ def __init__(
         supported_language: str = "en",
         version: str = "0.0.1",
         context: Optional[List[str]] = None,
+        analyzer_utils: Optional[PresidioAnalyzerUtils] = None,
```
please add `analyzer_utils` to the docstring

omri374 commented on presidio-analyzer/presidio_analyzer/predefined_recognizers/cfi_recognizer.py:
```diff
@@ -0,0 +1,204 @@
+            r"|DA|DN|DD|DM|DY|RA|RS|RP|RW|RF|RD|RM|OC|OP|OM|FF|FC|SR|ST|SE|SC|SF|SM|HR"
+            r"|HT|HE|HC|HF|HM|IF|IT|JE|JF|JC|JR|JT|KR|KT|KE|KC|KF|KY|KM|LL"
+            r"|LR|LS|TC|TT|TR|TI|TB|TD|TM|MC|MM)[A-Z]{4}$\b",
+            0.50,
```
Please reduce the score to 0.2. This could have false positives

omri374 commented on presidio-analyzer/presidio_analyzer/predefined_recognizers/in_gstin_recognizer.py:
```diff
@@ -0,0 +1,161 @@
+        supported_language: str = "en",
+        supported_entity: str = "IN_GSTIN",
+        replacement_pairs: Optional[List[Tuple[str, str]]] = None,
+        analyzer_utils=PresidioAnalyzerUtils(),
```
```suggestion
        analyzer_utils: Optional[PresidioAnalyzerUtils] = None,
```
Mutable objects should not be a default value (https://betterprogramming.pub/the-problem-with-mutable-types-as-default-values-in-functions-in-python-81ca88ff4d91)

Also, since all recognizers now support an `analyzer_utils` parameter, why not pass it during the recognizers instantiation, and only create it if it's missing for this specific recognizer?

This is one line out of two in `RecognizerRegistry` which instantiates the recognizers: 
https://github.com/devopam/presidio/blob/7d6ee388b40baca87809c8d8101b5fed3516cee0/presidio-analyzer/presidio_analyzer/recognizer_registry.py#L140

omri374 commented on presidio-analyzer/tests/test_analyzer_utils.py:
```diff
@@ -1,6 +1,17 @@
 
+
+@pytest.fixture(scope="module")
+def recognizer():
```
Please rename this fixture to `presidio_analyzer_utils`. The name is confusing :)
Also, is this fixture used anywhere?

omri374 commented on presidio-analyzer/presidio_analyzer/predefined_recognizers/in_pan_recognizer.py:
```diff
@@ -26,17 +26,17 @@ class InPanRecognizer(PatternRecognizer):
             "PAN (High)",
-            r"\b([A-Za-z]{3}[AaBbCcFfGgHhJjLlPpTt]{1}[A-Za-z]{1}[0-9]{4}[A-Za-z]{1})\b",
+            r"\b([A-Z]{3}[ABCFGHJLPT]{1}[A-Z]{1}(?!0000)[0-9]{4}[A-Z]{1})\b",
             0.85,
```
Please reduce to 0.5. This is a strong regex, but could still produce false positives

omri374 commented on presidio-analyzer/presidio_analyzer/predefined_recognizers/in_pan_recognizer.py:
```diff
@@ -26,17 +26,17 @@ class InPanRecognizer(PatternRecognizer):
         ),
         Pattern(
             "PAN (Medium)",
-            r"\b([A-Za-z]{5}[0-9]{4}[A-Za-z]{1})\b",
```
reduce to 0.45

omri374 commented on presidio-analyzer/presidio_analyzer/predefined_recognizers/isin_recognizer.py:
```diff
@@ -0,0 +1,72 @@
+        supported_language: str = "en",
+        supported_entity: str = "ISIN_CODE",
+        replacement_pairs: Optional[List[Tuple[str, str]]] = None,
+        analyzer_utils=PresidioAnalyzerUtils(),
```
See comment on other recognizer

omri374 commented on presidio-analyzer/presidio_analyzer/predefined_recognizers/isin_recognizer.py:
```diff
@@ -0,0 +1,72 @@
+    :param patterns: List of patterns to be used by this recognizer
+    :param context: List of context words to increase confidence in detection
+    :param supported_language: Language this recognizer supports
+    :param supported_entity: The entity this recognizer can detect
```
please add analyzer_utils to docstring

omri374 commented on presidio-analyzer/presidio_analyzer/predefined_recognizers/in_gstin_recognizer.py:
```diff
@@ -0,0 +1,161 @@
+    :param supported_language: Language this recognizer supports
+    :param supported_entity: The entity this recognizer can detect
+    :param replacement_pairs: List of tuples with potential replacement values
+    for different strings to be used during pattern matching.
```
please add analyzer_utils to docstring

omri374 commented on presidio-analyzer/presidio_analyzer/predefined_recognizers/in_gstin_recognizer.py:
```diff
@@ -0,0 +1,161 @@
+from presidio_analyzer.analyzer_utils import PresidioAnalyzerUtils
+
+
+# from presidio_analyzer.analyzer_utils import PresidioAnalyzerUtils as Utils
```
please remove

omri374 commented on presidio-analyzer/tests/test_analyzer_utils.py:
```diff
@@ -9,6 +20,12 @@
+luhn_mod_n_test_set = [
+    ["27AAACM6094R1ZP", True],
+    ["36AAICA3369H1ZJ", True],
+    ["36AAHAA2262Q1ZF", True],
```
Can we please add negative examples?

omri374 reviewed (commented):
Hi @devopam, added a few additional comments. This looks much better!
Thanks

omri374 commented:
Closing for now, please feel free to re-open if you'd like to continue working on this.
